keras_tuner/tuners/gridsearch.py

```diff
@@ -0,0 +1,218 @@
+            last_values = last_trial.hyperparameters.values
+            values = self._get_next_combination(last_values)
+        else:
+            # Use all default values for the first trial.
+            values = {hp.name: hp.default for hp in self.get_space().space}
+        if values is None:
+            return {"status": trial_module.TrialStatus.STOPPED, "values": None}
+        return {"status": trial_module.TrialStatus.RUNNING, "values": values}
+
+    def _get_next_combination(self, values):
+        """Get the next value combination.
+
+        This function is called on the second and later trials, but not on the
+        first one. Even new hps appear during search, the all hps in
+        self.get_space() should have corresponding values in the `values` arg.
+        This is because `values` is from the last trial, whose values has been
+        updated during that trial, which contains all seen hps.
+
+        We treat the default value of a hp as the first value, and the rest are
+        in normal order. We will enumerate the values according to this order.
+        Get all possible hyperparameter values
+        """
+
+        hps = self.get_space()
+        all_values = {}
+        for hp in hps.space:
+            value_list = list(hp.values)
+            if hp.default in value_list:
+                value_list.remove(hp.default)
+            # Put the default value first.
+            all_values[hp.name] = [hp.default] + value_list
+        default_values = {hp.name: hp.default for hp in hps.space}
+        new_values = copy.deepcopy(values)
+        hps.values = new_values
+
+        bumped_value = False
+
+        # Iterate in reverse order so that we can change the value under
+        # conditional scope first instead of change the condition value first.
+        for hp in reversed(hps.space):
+            name = hp.name
+            value = new_values[name]
+            # Bump up the hp value if possible and active.
+            if value != all_values[name][-1] and hps.is_active(hp):
+                index = all_values[name].index(value) + 1
+                new_values[name] = all_values[name][index]
+                bumped_value = True
+                break
+            # Otherwise, reset to its first value.
+            new_values[name] = default_values[name]
+
+        if bumped_value:
+            return new_values
+
+        return None
+
+
+class GridSearch(tuner_module.Tuner):
+    """The Grid search tuner.
+
+    It will try all the possible hyperparameter
+    combinations up to exhaustion.
+
+    For example:
+
+    ```py
+    optimizer = hp.Choice("model_name", values=["sgd", "adam"])
+    lr = hp.Choice("lr", values=[0.01, 0.1])
+    ```
+
+    This tuner will fit models for: ["sgd", 0.01], ["sgd", 0.1], ["adam", 0.01]
+    ["adam", 0.1].
+
+    For the following hyperparameter types, GridSearch will not exhaust all
+    possible values, but pick 10 samples evenly.
```

Comment:
Why 10? It seems this should be configurable.

Reply:
Revised the description for how to configure the granularity for Float and Int.